src/dataplane-protocol/src/record.rs

```diff
@@ -155,26 +155,24 @@ impl Encoder for DefaultAsyncBuffer {
 }
 
 impl Decoder for DefaultAsyncBuffer {
-    fn decode<T>(&mut self, src: &mut T, version: Version) -> Result<(), Error>
+    fn decode<T>(&mut self, src: &mut T, _: Version) -> Result<(), Error>
     where
         T: Buf,
     {
         trace!("decoding default asyncbuffer");
-        if let Some(ref mut buffer) = Arc::get_mut(&mut self.0) {
-            let mut len: i64 = 0;
-            len.decode_varint(src)?;
-            for _ in 0..len {
-                let mut value = <u8>::default();
-                value.decode(src, version)?;
-                buffer.push(value);
-            }
-            Ok(())
-        } else {
-            Err(Error::new(
-                ErrorKind::Other,
-                "Can't decode buffer while cloning".to_string(),
-            ))
-        }
+
+        let mut len: i64 = 0;
+        len.decode_varint(src)?;
+        let len = len as usize;
+
+        // Take `len` bytes from `src` and put them into a new BytesMut buffer
+        let slice = src.take(len);
+        let mut bytes_mut = bytes::BytesMut::with_capacity(len);
+        bytes_mut.put(slice);
+
+        // Replace the inner Bytes buffer of this DefaultAsyncBuffer
+        self.0 = bytes_mut.freeze();
```

Comment:
@sehz this is a section where I would like some careful review when you get a chance